### DISCUSSION TITLE
Run Avalonia dispatcher synchronously causes deadlock

### DISCUSSION THREAD

**Discussion by teocomi** [AvaloniaUI/Avalonia#8782] [Q&A] [answered]
We are loading Avalonia from inside a host application that is single-threaded.
The host app needs to perform a long-running loop (on the main thread) and we want to update the Avalonia UI in the meantime (progress bars & such), unfortunately, we cannot use async methods because of limitations in the host app and as a result, we have not yet found a way to update the Avalonia UI.

Things we tried:

`
Avalonia.Threading.Dispatcher.UIThread.InvokeAsync(() => progress.Update(conversionProgressDict), Avalonia.Threading.DispatcherPriority.Normal).Wait(); 
` => Deadlock

`Task.Run(async () => await Avalonia.Threading.Dispatcher.UIThread.InvokeAsync(() => progress.Update(conversionProgressDict), Avalonia.Threading.DispatcherPriority.Normal)).Wait();` => Deadlock

Also tried [this](https://stackoverflow.com/a/28307965) => Deadlock

The interesting bit is that some of our old WPF code using the `System.Windows.Threading.Dispatcher` works well as it runs synchronously.

Is there a way to call the Avalonia dispatcher synchronously without causing a deadlock?

I understand our configuration is a bit unusual but any help is appreciated!

**Comment by kekekeks** [comment, marked as answer]
Are you sure that you are always calling `Wait` from non-UI thread? Could you show the stack traces when deadlock happens?

**Reply by teocomi** [reply]
I'm actually calling `Wait` **from** the UI Thread, this is because the long-running loop is happening there and cannot be in another thread...
Sounds like the `Wait` is the reason for the deadlock, but what could I do otherwise? Without any dispatcher everything freezes....

Here is the call stack:

![image](https://user-images.githubusercontent.com/2679513/185594686-d1cca118-0ab3-4a5b-87be-4e32761d7838.png)

**Reply by kekekeks** [reply]
Well, you are asking the dispatcher to _schedule_ a callback on the UI thread and then blocking the UI thread, so dispatcher never gets a chance to execute the callback. What are you trying to achieve by waiting from the UI thread?

**Reply by teocomi** [reply]
My (naive) hope was to get the UI to update, without blocking it.

Funnily enough, we have just found that the following, using the Windows Dispatcher, actually does the trick:

```
  System.Windows.Threading.Dispatcher.CurrentDispatcher.Invoke(() =>
  {
    progress.Update(conversionProgressDict);
  }, System.Windows.Threading.DispatcherPriority.Background);

```

I'm not really sure why, but it seems to be purely down to our specific configuration...?

Thanks for your help!

**Reply by timunie** [reply]
If you don't want to block the UI you need to use await Task.Run() 

I'm loading huge data from DB like that and it works. You may want to show a progess indicator while doing so

**Reply by kekekeks** [reply]
So you basically need [DoEvents](https://docs.microsoft.com/en-us/office/vba/language/reference/user-interface-help/doevents-function) from VB6. 
I think on Windows and Linux you could achieve that by doing something like that:
```
var s = new CancellationTokenSource();
DispatcherTimer.RunOnce(() => s.Cancel(), TimeSpan.FromMilliseconds(10));
Dispatcher.UIThread.MainLoop(s.Token);
```

Which would pump the event loop for 10ms and return.

**Reply by teocomi** [reply]
This works too! Thanks a bunch 🥳
